django/contrib/auth/mixins.py

```diff
@@ -0,0 +1,89 @@
+from django.conf import settings
+from django.contrib.auth import REDIRECT_FIELD_NAME
+from django.contrib.auth.views import redirect_to_login
+from django.core.exceptions import ImproperlyConfigured, PermissionDenied
+from django.utils import six
+from django.utils.encoding import force_text
+
+
+class UserPassesTestMixin(object):
+    """
+    Mixin for class-based views that checks that the user passes the given by
+    the class method `test_func`.
+    """
+    login_url = None
+    permission_denied_message = ''
+    raise_exception = False
+    redirect_field_name = REDIRECT_FIELD_NAME
+
+    def dispatch(self, request, *args, **kwargs):
+        user_test_result = self.get_test_func()(request.user)
+
+        if not user_test_result:
+            return self.handle_no_permission(request)
+
+        return super(UserPassesTestMixin, self).dispatch(request, *args, **kwargs)
+
+    def get_test_func(self):
+        return getattr(self, "test_func")
+
+    def test_func(self, user):
+        raise NotImplementedError(
+            '{0} is missing implementation of the '
+            'test_func method.'.format(self.__class__.__name__)
+        )
+
+    def get_login_url(self):
+        """
+        Override this method to customize the login_url.
+        """
+        login_url = self.login_url or settings.LOGIN_URL
+        if not login_url:
+            raise ImproperlyConfigured(
+                'Define {0}.login_url or settings.LOGIN_URL or override '
+                '{0}.get_login_url().'.format(self.__class__.__name__)
+            )
+        return force_text(login_url)
+
+    def get_permission_denied_message(self):
+        return self.permission_denied_message
+
+    def get_redirect_field_name(self):
+        """
+        Override this method to customize the redirect_field_name.
+        """
+        return self.redirect_field_name
+
+    def handle_no_permission(self, request):
+        # In case the 403 handler should be called, raise the exception.
+        if self.raise_exception:
+            raise PermissionDenied(self.get_permission_denied_message())
+        return redirect_to_login(request.get_full_path(), self.get_login_url(), self.get_redirect_field_name())
+
+
+class LoginRequiredMixin(UserPassesTestMixin):
+    """
+    Mixin for class-based views that checks that the user is logged in,
+    redirecting to the login page if necessary.
+    """
+
+    def test_func(self, user):
```

Comment:
One thing we can't do with the current implement, is joining `LoginRequiredMixin` and `PermissionRequiredMixin` in a single view: `class MyView(LoginRequiredMixin, PermissionRequiredMixin, View)`

On June 9, 2015 7:20:21 PM GMT+02:00, Raphael Michel <notifications@github.com> wrote:
>> +        return self.redirect_field_name
>> +
>> +    def handle_no_permission(self, request):
>> +        # In case the 403 handler should be called, raise the
>exception.
>> +        if self.raise_exception:
>> +            raise
>PermissionDenied(self.get_permission_denied_message())
>> +        return redirect_to_login(request.get_full_path(),
>self.get_login_url(), self.get_redirect_field_name())
>> +
>> +
>> +class LoginRequiredMixin(UserPassesTestMixin):
>> +    """
>> +    Mixin for class-based views that checks that the user is logged
>in,
>> +    redirecting to the login page if necessary.
>> +    """
>> +
>> +    def test_func(self, user):
>
>I have to admit that I don't really understand what your point is.
>
>* The ``UserPassesTestMixin`` in this pull request should behave
>exactly the same way than the one in django-braces does. The only
>differences are:
>  * ``raise_exception`` can only be a boolean value
>  * ``redirect_unauthenticated_users`` does not exist
>* The other two mixins use ``UserPassesTestMixin`` for the
>functionality (just like the decorators) instead of all three using a
>fourth class as their parent.
>
>I see that the django-braces ones are more powerful, but I don't yet
>see a use case where the ones in this PR are harder or more complex to
>use/override. I actually feel that the ones in this PR have fewer parts
>to deal with. Feel free to tell me that I'm wrong :)
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/4836/files#r32038836